src/Cake.Common.Tests/Unit/Tools/VSWhere/Product/VSWhereProductTests.cs

```diff
@@ -0,0 +1,188 @@
+            [InlineData("./tools/vswhere/vswhere.exe", "/Working/tools/vswhere/vswhere.exe")]
+            public void Should_Use_VSWhere_Executable_From_Tool_Path_If_Provided(string toolPath, string expected)
+            {
+                // Given
+                var fixture = new VSWhereProductFixture();
+                fixture.Settings.ToolPath = toolPath;
+                fixture.GivenSettingsToolPathExist();
+
+                // When
+                var result = fixture.Run();
+
+                // Then
+                Assert.Equal(expected, result.Path.FullPath);
+            }
+
+            [WindowsTheory]
+            [InlineData("C:/vswhere/vswhere.exe", "C:/vswhere/vswhere.exe")]
+            public void Should_Use_VSWhere_Executable_From_Tool_Path_If_Provided_On_Windows(string toolPath, string expected)
+            {
+                // Given
+                var fixture = new VSWhereProductFixture();
+                fixture.Settings.ToolPath = toolPath;
+                fixture.GivenSettingsToolPathExist();
+
+                // When
+                var result = fixture.Run();
+
+                // Then
+                Assert.Equal(expected, result.Path.FullPath);
+            }
+
+            [Fact]
+            public void Should_Throw_If_Process_Was_Not_Started()
+            {
+                // Given
+                var fixture = new VSWhereProductFixture();
+                fixture.GivenProcessCannotStart();
+
+                // When
+                var result = Record.Exception(() => fixture.Run());
+
+                // Then
+                Assert.IsCakeException(result, "VSWhere: Process was not started.");
+            }
+
+            [Fact]
+            public void Should_Throw_If_Process_Has_A_Non_Zero_Exit_Code()
+            {
+                // Given
+                var fixture = new VSWhereProductFixture();
+                fixture.GivenProcessExitsWithCode(1);
+
+                // When
+                var result = Record.Exception(() => fixture.Run());
+
+                // Then
+                Assert.IsCakeException(result, "VSWhere: Process returned an error (exit code 1).");
+            }
+
+            [Fact]
+            public void Should_Find_VSWhere_Executable_If_Tool_Path_Not_Provided()
+            {
+                // Given
+                var fixture = new VSWhereProductFixture();
+
+                // When
+                var result = fixture.Run();
+
+                // Then
+                Assert.Equal("/Working/tools/vswhere.exe", result.Path.FullPath);
+            }
+
+            [Fact]
+            public void Should_Add_Mandatory_Arguments()
+            {
+                // Given
+                var fixture = new VSWhereProductFixture();
+
+                // When
+                var result = fixture.Run();
+
+                // Then
+                Assert.Equal("-products \"BuildTools\" -property installationPath -nologo", result.Args);
```

Comment:
While this is fine for tests, keep in mind the full product name is "Microsoft.VisualStudio.Product.BuildTools". I have a PR to support short names but haven't merged it yet because of no demand. It would also increase the size of the EXE a bit more.